File: harness/lib/observability/node/health.ts
interface HealthCheck {
  name: string;
  check(): Promise<void>;
}

interface HealthStatus {
  status: 'ok' | 'degraded' | 'down';
  checks: Record<string, { status: 'ok' | 'error'; error?: string }>;
  timestamp: string;
}

const checks: HealthCheck[] = [];

export function registerHealthCheck(check: HealthCheck): void {
  checks.push(check);
}

async function runHealthChecks(): Promise<HealthStatus> {
  const results: HealthStatus['checks'] = {};
  let overallStatus: HealthStatus['status'] = 'ok';

  await Promise.allSettled(
    checks.map(async (check) => {
      try {
        await check.check();
        results[check.name] = { status: 'ok' };
      } catch (err) {
        results[check.name] = {
          status: 'error',
          error: err instanceof Error ? err.message : String(err),
        };
        overallStatus = 'down';
      }
    }),
  );

  return {
    status: overallStatus,
    checks: results,
    timestamp: new Date().toISOString(),
  };
}

/**
 * Cria um Express Router com endpoints /health/live e /health/ready.
 *
 * Uso:
 *   import express from 'express';
 *   import { createHealthRouter, registerHealthCheck } from '@harness/observability';
 *
 *   registerHealthCheck({
 *     name: 'database',
 *     check: async () => { await db.raw('SELECT 1') },
 *   });
 *
 *   app.use(createHealthRouter());
 */
export function createHealthRouter() {
  // Retorna um objeto compatível com Express Router sem importar Express diretamente
  const routes: Array<{ path: string; handler: (req: unknown, res: { json: (data: unknown) => void; status: (code: number) => { json: (data: unknown) => void } }) => void }> = [];

  routes.push({
    path: '/health/live',
    handler: (_req, res) => {
      res.json({ status: 'ok', timestamp: new Date().toISOString() });
    },
  });

  routes.push({
    path: '/health/ready',
    handler: async (_req, res) => {
      const result = await runHealthChecks();
      const statusCode = result.status === 'ok' ? 200 : 503;
      res.status(statusCode).json(result);
    },
  });

  return routes;
}

File: harness/lib/observability/python/health.py
"""
Health endpoints /health/live e /health/ready para FastAPI.
"""
from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Callable, Optional

HealthCheckFn = Callable[[], Any]
_checks: list[tuple[str, HealthCheckFn]] = []


def register_health_check(name: str, check: HealthCheckFn) -> None:
    _checks.append((name, check))


async def run_health_checks() -> dict[str, Any]:
    results: dict[str, Any] = {}
    overall = "ok"

    async def run_one(name: str, check: HealthCheckFn) -> None:
        nonlocal overall
        try:
            result = check()
            if asyncio.iscoroutine(result):
                await result
            results[name] = {"status": "ok"}
        except Exception as e:
            results[name] = {"status": "error", "error": str(e)}
            overall = "down"

    await asyncio.gather(*[run_one(n, c) for n, c in _checks])
    return {
        "status": overall,
        "checks": results,
        "timestamp": datetime.now(tz=timezone.utc).isoformat(),
    }


try:
    from fastapi import APIRouter
    from fastapi.responses import JSONResponse

    health_router = APIRouter()

    @health_router.get("/health/live")
    async def liveness() -> dict[str, str]:
        return {"status": "ok", "timestamp": datetime.now(tz=timezone.utc).isoformat()}

    @health_router.get("/health/ready")
    async def readiness() -> JSONResponse:
        result = await run_health_checks()
        status_code = 200 if result["status"] == "ok" else 503
        return JSONResponse(content=result, status_code=status_code)

except ImportError:
    pass
